OMCompiler/SimulationRuntime/c/simulation/solver/jacobianSymbolical.c

```diff
@@ -0,0 +1,88 @@
+ * RECIPIENT'S ACCEPTANCE OF THE OSMC PUBLIC LICENSE OR THE GPL VERSION 3,
+ * ACCORDING TO RECIPIENTS CHOICE.
+ *
+ * The OpenModelica software and the OSMC (Open Source Modelica Consortium)
+ * Public License (OSMC-PL) are obtained from OSMC, either from the above
+ * address, from the URLs: http://www.openmodelica.org or
+ * http://www.ida.liu.se/projects/OpenModelica, and in the OpenModelica
+ * distribution. GNU version 3 is obtained from:
+ * http://www.gnu.org/copyleft/gpl.html. The New BSD License is obtained from:
+ * http://www.opensource.org/licenses/BSD-3-Clause.
+ *
+ * This program is distributed WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, EXCEPT AS
+ * EXPRESSLY SET FORTH IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE
+ * CONDITIONS OF OSMC-PL.
+ *
+ */
+
+ /*! \file jacobian_symbolical.c
+ */
+
+#include "simulation/solver/jacobianSymbolical.h"
+
+
+/**
+ * \brief Generic parallel computation of the colored Jacobian.
+ *
+ * Exploiting coloring and sparse structure. Used from DASSL and IDA solvers.
+ * Only matrix storing format differs for them and therefore setJacElementFunc
+ * is used to access matrix A.
+ *
+ * \param [in]      rows                Number of rows of jacobian.
+ * \param [in]      columns             Number of columns of jacobian.
+ * \param [in]      spp                 Pointer to sparse pattern.
+ * \param [in/out]  matrixA             Internal data of solvers to store jacobian.
+ * \param [in]      jacColumns          Number of colors (=number of columns for compressed structure) of jacobian.
+ * \param [in]      data
+ * \param [in]      threadData
+ * \param [in]      setJacElementFunc   Function to set element (i,j) in matrix A.
+ */
+void genericColoredSymbolicJacobianEvaluation(int rows, int columns, SPARSE_PATTERN* spp,
+                                              void* matrixA, ANALYTIC_JACOBIAN* jacColumns, DATA* data,
+                                              threadData_t* threadData,
+                                              void (*setJacElement)(int, int, int, double, void*, int))
+{
+  ANALYTIC_JACOBIAN* t_jac = jacColumns;
+
+  unsigned int i, j, currentIndex, nth;
+  for (i=0; i < spp->maxColors; i++) {
+    /* Set seed vector for current color */
+    for (j=0; j < columns; j++) {
+      if (spp->colorCols[j]-1 == i) {
+        t_jac->seedVars[j] = 1;
+      }
+    }
+
+    /* Evaluate with updated seed vector */
+    data->callback->functionJacA_column(data, threadData, t_jac, NULL);
+    /* Save jacobian elements in matrixA*/
+    for (j=0; j < columns; j++) {
+      if (t_jac->seedVars[j] == 1) {
+        nth = spp->leadindex[j];
+        while (nth < spp->leadindex[j+1]) {
+          currentIndex = spp->index[nth];
+          (*setJacElement)(currentIndex, j, nth, t_jac->resultVars[currentIndex], matrixA, rows);
+          nth++;
+        };
```

Comment:
Nitpicking 🥇 If you have time please remove this semicolon

Reply:
Done